### PR TITLE
Throw if unique_ptr or array allocation fails due to SafeInt overflow

### PR DESCRIPTION
**Pull request by skottmckay**
### Description

If we fail to calculate the buffer size (due to overflow) we currently return a nullptr. This is inconsistent as an actual memory allocation failure throws. An overflow would typically be due to bad input so an exception makes more sense given that.

Change to throw so code using MakeUniquePtr* and AllocArray* doesn't need to check for nullptr.

Add some extra info to the log message to help debugging.

### Motivation and Context

Should help with #18905 by avoiding the invalid attempted usage of a nullptr from the allocation. Extra info _might_ help with figuring out where the overflow is coming from which is the real issue.

### CODE REVIEW

File: include/onnxruntime/core/framework/allocator.h [hunk 1]
```diff
@@ -150,7 +156,8 @@ class IAllocator {
   static IAllocatorUniquePtr<T> MakeUniquePtr(std::shared_ptr<IAllocator> allocator, size_t count_or_bytes,
                                               bool use_reserve = false,
                                               Stream* stream = nullptr, WaitNotificationFn wait_fn = nullptr) {
-    if (allocator == nullptr) return nullptr;
+    assert(allocator != nullptr);
```
Comment (edgchen1) [inline include/onnxruntime/core/framework/allocator.h:159]: nit: ORT_ENFORCE? I think an exception is preferable to a segfault in a release build.

Comment (skottmckay) [reply include/onnxruntime/core/framework/allocator.h:159]: I started with ORT_ENFORCE and changed it as passing a null allocator should never happen, and must not currently as code using this isn't in general checking for a nullptr return.  

The ORT_ENFORCE adds extra binary size as this is a templatized function.

Comment (skottmckay) [reply include/onnxruntime/core/framework/allocator.h:159]: I'll do some tests to try and minimize the binary size impact. Can try splitting out the validation here and for the CalcMemSizeForArray call so those aren't templatized on the allocation type.

File: include/onnxruntime/core/framework/allocator.h [hunk 2]
```diff
@@ -161,38 +168,60 @@ class IAllocator {
     if constexpr (!std::is_void<T>::value) {
       // sizeof(void) isn't valid, but the compiler isn't smart enough to ignore that this line isn't
       // reachable if T is void. use std::conditional to 'use' void* in the sizeof call
-      if (!CalcMemSizeForArray(
-              count_or_bytes, sizeof(typename std::conditional<std::is_void<T>::value, void*, T>::type), &alloc_size)) {
-        return nullptr;
-      }
+      const auto size = sizeof(typename std::conditional<std::is_void<T>::value, void*, T>::type);
```
Comment (edgchen1) [inline include/onnxruntime/core/framework/allocator.h:171]:
```suggestion
      constexpr auto size = sizeof(typename std::conditional<std::is_void<T>::value, void*, T>::type);
```

File: include/onnxruntime/core/framework/allocator.h [hunk 3]
```diff
@@ -8,6 +8,7 @@
 #include "core/session/onnxruntime_c_api.h"
 #include "ortdevice.h"
 #include "ortmemoryinfo.h"
+#include <cassert>
```
Comment (edgchen1) [inline include/onnxruntime/core/framework/allocator.h:11]: do we still need `<cassert>`?

**Review by edgchen1 — APPROVED** [review]

File: include/onnxruntime/core/framework/allocator.h [hunk 4]
```diff
@@ -144,13 +150,14 @@ class IAllocator {
      @param stream Which stream instance allocated chunk will be used with.
      @param wait_fn If the allocator want to dynamic reuse a chunk from another stream, use this wait_fn to sync on
                     the target stream to make the reuse safe.
-     @returns std::unique_ptr with allocated memory and deleter.
+     @returns std::unique_ptr with allocated memory and deleter. Throws if it cannot allocate memory.
```
Comment (edgchen1) [inline include/onnxruntime/core/framework/allocator.h:153]: > Throws if it cannot allocate memory.

Does this mean that IAllocator::Alloc is expected to throw and not return nullptr on error? Looks like the CPU and CUDA EP allocators already do that anyway.

Could document that assumption or enforce that `p != nullptr || count_or_bytes == 0`.

**Pull request merged**